### PR TITLE
virt-launcher: Use libvirt function to set SSH keys in the guest

### PR DESCRIPTION
**Pull request by akrejcir**
**What this PR does / why we need it**:
This PR simplifies how SSH keys are set in the guest. It uses functions provided by libvirt, instead of writing to the `authorized_keys` file directly.

**Release note**:
```release-note
None
```

### CODE REVIEW

File: pkg/virt-launcher/virtwrap/access-credentials/access_credentials.go [hunk 1]
```diff
@@ -597,3 +498,86 @@ func (l *AccessCredentialManager) HandleQemuAgentAccessCredentials(vmi *v1.Virtu
...
+	// filepath mapped to secretNames
+	userSSHMap map[string][]string
+	// maps users to passwords
+	userPasswordMap map[string]string
+}
+
+func (a *accessCredentialsInfo) addAccessCredential(accessCred *v1.AccessCredential) error {
+	secretName := getSecret(accessCred)
+	if secretName == "" {
+		return nil
```
Comment (jcanocan) [inline pkg/virt-launcher/virtwrap/access-credentials/access_credentials.go:514]: Should we here return an error? Or it is not consider a problematic situation to not have the secret?

**Review by jcanocan — COMMENTED** [review]
Nice work! I just have one small comment. IMHO I would like to have a link to the documentation pointing out from the user perspective, how to use this or an small yaml example.

File: pkg/virt-launcher/virtwrap/access-credentials/access_credentials.go (hunk 1 above)
Comment (akrejcir) [reply pkg/virt-launcher/virtwrap/access-credentials/access_credentials.go:560]: It's not an error. If user does not specify a secret, then there is no key to add to the accumulation struct.

Comment (jcanocan) [reply pkg/virt-launcher/virtwrap/access-credentials/access_credentials.go:560]: Got it! Thanks.

File: pkg/virt-launcher/virtwrap/access-credentials/access_credentials.go [hunk 2]
```diff
@@ -306,7 +309,25 @@ func (l *AccessCredentialManager) pingAgent(domName string) error {
...
+func (l *AccessCredentialManager) agentSetAuthorizedKeys(domName string, user string, authorizedKeys []string) (err error) {
+	domain, err := l.virConn.LookupDomainByName(domName)
+	if err != nil {
+		return err
+	}
+	defer domain.Free()
+
+	// Zero flags argument means that the authorized_keys file is overwritten with the authorizedKeys
+	err = domain.AuthorizedSSHKeysSet(user, authorizedKeys, 0)
+	if !errors.Is(err, libvirt.ERR_NO_SUPPORT) {
```
Comment (ksimon1) [inline pkg/virt-launcher/virtwrap/access-credentials/access_credentials.go:321]: what about use case when err == libvirt.ERR_NO_SUPPORT?
Does it has to be handled somehow?

File: pkg/virt-launcher/virtwrap/access-credentials/access_credentials.go [hunk 3]
```diff
@@ -597,3 +545,86 @@ func (l *AccessCredentialManager) HandleQemuAgentAccessCredentials(vmi *v1.Virtu
...
+	secretDir := getSecretDir(secretName)
+	files, err := os.ReadDir(secretDir)
+	if err != nil {
+		return fmt.Errorf("error encountered reading secrets file list from base directory %s: %w", secretDir, err)
+	}
+
+	if isSSHPublicKey(accessCred) {
+		for _, user := range accessCred.SSHPublicKey.PropagationMethod.QemuGuestAgent.Users {
+			secrets, ok := a.userSSHMap[user]
+			if !ok {
```
Comment (lyarwood) [inline pkg/virt-launcher/virtwrap/access-credentials/access_credentials.go:573]: nit as this isn't new code but you should always be able to `a.userSSHMap[user] = append(secrets, secretName)` here regardless of the value of `ok`?

https://go.dev/play/p/h-FeG2MHZ1N

File: pkg/virt-launcher/virtwrap/access-credentials/access_credentials.go [hunk 4]
```diff
@@ -597,3 +545,86 @@ func (l *AccessCredentialManager) HandleQemuAgentAccessCredentials(vmi *v1.Virtu
...
+func (a *accessCredentialsInfo) addAccessCredential(accessCred *v1.AccessCredential) error {
+	secretName := getSecret(accessCred)
+	if secretName == "" {
+		return nil
+	}
+
+	secretDir := getSecretDir(secretName)
+	files, err := os.ReadDir(secretDir)
+	if err != nil {
+		return fmt.Errorf("error encountered reading secrets file list from base directory %s: %w", secretDir, err)
```
Comment (codingben) [inline pkg/virt-launcher/virtwrap/access-credentials/access_credentials.go:567]: Can you please change it to `error occurred while reading the list of secrets files from the base directory %s: %w`?

File: pkg/virt-launcher/virtwrap/access-credentials/access_credentials.go (hunk 3 above)
Comment (akrejcir) [reply pkg/virt-launcher/virtwrap/access-credentials/access_credentials.go]: Right. I've simplified this.

File: pkg/virt-launcher/virtwrap/access-credentials/access_credentials.go (hunk 2 above)
Comment (akrejcir) [reply pkg/virt-launcher/virtwrap/access-credentials/access_credentials.go]: I missed a branch where `err == nil`. Fixed.

File: pkg/virt-launcher/virtwrap/access-credentials/access_credentials.go [hunk 5]
```diff
@@ -597,3 +546,81 @@ func (l *AccessCredentialManager) HandleQemuAgentAccessCredentials(vmi *v1.Virtu
...
+
+		var authorizedKeys []string
+		for _, file := range files {
+			if file.IsDir() || strings.HasPrefix(file.Name(), "..") {
+				continue
+			}
+
+			pubKeyBytes, err := os.ReadFile(filepath.Join(secretDir, file.Name()))
+			if err != nil {
+				return fmt.Errorf("error encountered readding access credential secret file [%s]: %w", filepath.Join(secretDir, file.Name()), err)
```
Comment (codingben) [inline pkg/virt-launcher/virtwrap/access-credentials/access_credentials.go:584]: Can you please change it to `error occurred while reading the access credential secret file [%s]: %w`?

File: pkg/virt-launcher/virtwrap/access-credentials/access_credentials.go [hunk 6]
```diff
@@ -597,3 +546,81 @@ func (l *AccessCredentialManager) HandleQemuAgentAccessCredentials(vmi *v1.Virtu
...
+
+	} else if isUserPassword(accessCred) {
+		for _, file := range files {
+			if file.IsDir() || strings.HasPrefix(file.Name(), "..") {
+				continue
+			}
+
+			passwordBytes, err := os.ReadFile(filepath.Join(secretDir, file.Name()))
+			if err != nil {
+				return fmt.Errorf("error encountered reading access credential secret file [%s]: %w", filepath.Join(secretDir, file.Name()), err)
```
Comment (codingben) [inline pkg/virt-launcher/virtwrap/access-credentials/access_credentials.go:606]: Can you please change it to `error occurred while reading the access credential secret file [%s]: %w`?

File: pkg/virt-launcher/virtwrap/access-credentials/access_credentials.go [hunk 7]
```diff
@@ -597,3 +545,86 @@ func (l *AccessCredentialManager) HandleQemuAgentAccessCredentials(vmi *v1.Virtu
...
+			if !ok {
+				a.userSSHMap[user] = []string{secretName}
+			} else {
+				a.userSSHMap[user] = append(secrets, secretName)
+			}
+		}
+
+		var authorizedKeys []string
+		for _, file := range files {
+			if file.IsDir() || strings.HasPrefix(file.Name(), "..") {
```
[the remark this answers is not in the document]
Comment (codingben) [reply pkg/virt-launcher/virtwrap/access-credentials/access_credentials.go:578]: I know that there is only one prefix usually, e.g. `.ssh` or something similar.

[the remark this answers is not in the document]
Comment (akrejcir) [reply pkg/virt-launcher/virtwrap/access-credentials/access_credentials.go:578]: When a `configMap` or `secret` is mounted there are directories prefixed with `..`. For example, this is how it looks like when a secret is mounted to pod:
```
$ ls -lah

drwxrwsrwt. 3 root qemu  100 May 23 10:50 .
drwxr-sr-x. 3 root qemu 4.0K May 23 10:50 ..
drwxr-sr-x. 2 root qemu   60 May 23 10:50 ..2023_05_23_10_50_30.2449578983
lrwxrwxrwx. 1 root qemu   32 May 23 10:50 ..data -> ..2023_05_23_10_50_30.2449578983
lrwxrwxrwx. 1 root qemu   16 May 23 10:50 myKey.pub -> ..data/myKey.pub
```

This symlink structure is probably used to atomically update all files when the `Secret` object is updated in the API server.

File: pkg/virt-launcher/virtwrap/access-credentials/access_credentials.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (akrejcir) [reply pkg/virt-launcher/virtwrap/access-credentials/access_credentials.go]: I've done it differently.

File: pkg/virt-launcher/virtwrap/access-credentials/access_credentials.go (hunk 4 above)
Comment (akrejcir) [reply pkg/virt-launcher/virtwrap/access-credentials/access_credentials.go]: Done

File: pkg/virt-launcher/virtwrap/access-credentials/access_credentials.go (hunk 5 above)
Comment (akrejcir) [reply pkg/virt-launcher/virtwrap/access-credentials/access_credentials.go]: Done

File: pkg/virt-launcher/virtwrap/access-credentials/access_credentials.go (hunk 6 above)
Comment (akrejcir) [reply pkg/virt-launcher/virtwrap/access-credentials/access_credentials.go]: Done

File: pkg/virt-launcher/virtwrap/access-credentials/access_credentials.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (akrejcir) [reply pkg/virt-launcher/virtwrap/access-credentials/access_credentials.go:578]: Added a comment.

**Review by lyarwood — APPROVED** [review]
/lgtm

**Comment by akrejcir** [comment]
CI passed. 
Can someone approve this?

**Review by xpivarc — CHANGES_REQUESTED** [review]

File: pkg/virt-launcher/virtwrap/access-credentials/access_credentials_test.go [hunk 8]
```diff
@@ -116,98 +116,14 @@ var _ = Describe("AccessCredentials", func() {
...
-		mergedKeys := base64.StdEncoding.EncodeToString([]byte(authorizedKeys))
-		expectedWriteCmd := fmt.Sprintf(`{"execute": "guest-file-write", "arguments": { "handle": 1000, "buf-b64": "%s" } }`, mergedKeys)
-
-		expectedCloseCmd := `{"execute": "guest-file-close", "arguments": { "handle": 1000 } }`
-
-		expectedExecReturn := `{"return":{"pid":789}}`
-		expectedStatusCmd := `{"execute": "guest-exec-status", "arguments": { "pid": 789 } }`
+		mockConn.EXPECT().LookupDomainByName(domName).Return(mockDomain, nil).AnyTimes()
+		mockDomain.EXPECT().AuthorizedSSHKeysSet(user, authorizedKeys, gomock.Any()).Return(nil).AnyTimes()
+		mockDomain.EXPECT().Free().AnyTimes()
```
[the remark this answers is not in the document]
Comment (akrejcir) [reply pkg/virt-launcher/virtwrap/access-credentials/access_credentials_test.go:124]: Can you give me more information why not? I assumed it's irrelevant how many times is the mock called.

[the remark this answers is not in the document]
Comment (xpivarc) [reply pkg/virt-launcher/virtwrap/access-credentials/access_credentials_test.go:124]: I am more interested in how many times the `Free` is called as that can prevent us from leaks. There were problems in the past.

[the remark this answers is not in the document]
Comment (akrejcir) [reply pkg/virt-launcher/virtwrap/access-credentials/access_credentials_test.go]: Makes sense.
Done.

File: pkg/virt-launcher/virtwrap/access-credentials/access_credentials.go [hunk 9]
```diff
@@ -306,7 +309,26 @@ func (l *AccessCredentialManager) pingAgent(domName string) error {
...
+	}()
+	if errors.Is(err, libvirt.ERR_NO_SUPPORT) {
+		// If AuthorizedSSHKeysSet method is not supported, use the old method
+		desiredAuthorizedKeys := strings.Join(authorizedKeys, "\n")
+		return l.agentWriteAuthorizedKeysFile(domName, user, desiredAuthorizedKeys)
+	}
+	return err
+}
+
+func (l *AccessCredentialManager) agentWriteAuthorizedKeysFile(domName string, user string, desiredAuthorizedKeys string) (err error) {
```
Comment (xpivarc) [inline pkg/virt-launcher/virtwrap/access-credentials/access_credentials.go:331]: Maybe you want to mark this as deprecated and/or also include "legacy" or a similar word in the name.

File: pkg/virt-launcher/virtwrap/access-credentials/access_credentials.go [hunk 10]
```diff
@@ -306,7 +309,26 @@ func (l *AccessCredentialManager) pingAgent(domName string) error {
...
+		domain, err := l.virConn.LookupDomainByName(domName)
+		if err != nil {
+			return err
+		}
+		defer domain.Free()
+
+		// Zero flags argument means that the authorized_keys file is overwritten with the authorizedKeys
+		return domain.AuthorizedSSHKeysSet(user, authorizedKeys, 0)
+	}()
+	if errors.Is(err, libvirt.ERR_NO_SUPPORT) {
```
[the remark this answers is not in the document]
Comment (akrejcir) [reply pkg/virt-launcher/virtwrap/access-credentials/access_credentials.go:323]: I think, if an older version of libvirt is used that does not have the `AuthorizedSSHKeysSet` function.
It's mentioned in the documentation for the library:
https://github.com/kubevirt/kubevirt/blob/065f00dc89a24eea9b9e734a5cd7ebd43cd9f8d0/vendor/libvirt.org/go/libvirt/doc.go#L31

[the remark this answers is not in the document]
Comment (xpivarc) [reply pkg/virt-launcher/virtwrap/access-credentials/access_credentials.go:323]: Ok, then it's as I thought. You can simply assume we have the right version imho. We do not make any guarantees we work with older versions.
@vladikr keep me honest

File: pkg/virt-launcher/virtwrap/access-credentials/access_credentials.go (hunk 9 above)
Comment (akrejcir) [reply pkg/virt-launcher/virtwrap/access-credentials/access_credentials.go:331]: If I can assume that we use correct libvirt version, I can remove this function.

**Pull request merged**